File: python/tests/api/expressions/test_misc_literal_conversions.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from decimal import Decimal
import sys
import unittest
import uuid


from iceberg.api.expressions import Literal
from iceberg.api.types import (BinaryType,
                               BooleanType,
                               DateType,
                               DecimalType,
                               DoubleType,
                               FixedType,
                               FloatType,
                               IntegerType,
                               LongType,
                               StringType,
                               TimestampType,
                               TimeType,
                               UUIDType)


class TestMiscLiteralConversions(unittest.TestCase):

    def test_identity_conversions(self):
        pairs = [(Literal.of(True), BooleanType.get()),
                 (Literal.of(34), IntegerType.get()),
                 (Literal.of(3400000000), LongType.get()),
                 (Literal.of(34.11), FloatType.get()),
                 (Literal.of(34.11), DoubleType.get()),
                 (Literal.of(Decimal(34.55).quantize(Decimal("0.01"))), DecimalType.of(9, 2)),
                 (Literal.of("2017-08-18"), DateType.get()),
                 (Literal.of("14:21:01.919"), TimeType.get()),
                 (Literal.of("2017-08-18T14:21:01.919"), TimestampType.without_timezone()),
                 (Literal.of("abc"), StringType.get()),
                 (Literal.of(uuid.uuid4()), UUIDType.get())
                 ]

        if sys.version_info[0] >= 3:
            pairs = pairs + [(Literal.of(bytes([0x01, 0x02, 0x03])), FixedType.of_length(3)),
                             (Literal.of(bytearray([0x03, 0x04, 0x05, 0x06])), BinaryType.get())]

        for pair in pairs:
            expected = pair[0].to(pair[1])
            self.assertIs(expected, expected.to(pair[1]))

    def test_binary_to_fixed(self):
        if sys.version_info[0] >= 3:
            lit = Literal.of(bytearray([0x00, 0x01, 0x02]))
            fixed_lit = lit.to(FixedType.of_length(3))
            self.assertIsNotNone(fixed_lit)
            self.assertEqual(lit.value, fixed_lit.value)
            self.assertIsNone(lit.to(FixedType.of_length(4)))
            self.assertIsNone(lit.to(FixedType.of_length(2)))

    def test_fixed_to_binary(self):
        if sys.version_info[0] >= 3:
            lit = Literal.of(bytes([0x00, 0x01, 0x02]))
            binary_lit = lit.to(BinaryType.get())
            self.assertIsNotNone(binary_lit)
            self.assertEqual(lit.value, binary_lit.value)

    def test_invalid_boolean_conversions(self):
        self.assert_invalid_conversions(Literal.of(True), [IntegerType.get(),
                                                           LongType.get(),
                                                           FloatType.get(),
                                                           DoubleType.get(),
                                                           DateType.get(),
                                                           TimeType.get(),
                                                           TimestampType.with_timezone(),
                                                           TimestampType.without_timezone(),
                                                           DecimalType.of(9, 2),
                                                           StringType.get(),
                                                           UUIDType.get(),
                                                           FixedType.of_length(1),
                                                           BinaryType.get()])

    def test_invalid_integer_conversions(self):
        self.assert_invalid_conversions(Literal.of(34), [BooleanType.get(),
                                                         TimeType.get(),
                                                         TimestampType.with_timezone(),
                                                         TimestampType.without_timezone(),
                                                         StringType.get(),
                                                         UUIDType.get(),
                                                         FixedType.of_length(1),
                                                         BinaryType.get()])

    def test_invalid_long_conversions(self):
        self.assert_invalid_conversions(Literal.of(34).to(LongType.get()), [BooleanType.get(),
                                                                            DateType.get(),
                                                                            StringType.get(),
                                                                            UUIDType.get(),
                                                                            FixedType.of_length(1),
                                                                            BinaryType.get()])

    def test_invalid_float_conversions(self):
        self.assert_invalid_conversions(Literal.of(34.11), [BooleanType.get(),
                                                            IntegerType.get(),
                                                            LongType.get(),
                                                            DateType.get(),
                                                            TimeType.get(),
                                                            TimestampType.with_timezone(),
                                                            TimestampType.without_timezone(),
                                                            StringType.get(),
                                                            UUIDType.get(),
                                                            FixedType.of_length(1),
                                                            BinaryType.get()])

    def test_invalid_double_conversions(self):
        self.assert_invalid_conversions(Literal.of(34.11).to(DoubleType.get()), [BooleanType.get(),
                                                                                 IntegerType.get(),
                                                                                 LongType.get(),
                                                                                 DateType.get(),
                                                                                 TimeType.get(),
                                                                                 TimestampType.with_timezone(),
                                                                                 TimestampType.without_timezone(),
                                                                                 StringType.get(),
                                                                                 UUIDType.get(),
                                                                                 FixedType.of_length(1),
                                                                                 BinaryType.get()])

    def test_invalid_date_conversions(self):
        self.assert_invalid_conversions(Literal.of("2017-08-18").to(DateType.get()), [BooleanType.get(),
                                                                                      IntegerType.get(),
                                                                                      LongType.get(),
                                                                                      FloatType.get(),
                                                                                      DoubleType.get(),
                                                                                      TimeType.get(),
                                                                                      TimestampType.with_timezone(),
                                                                                      TimestampType.without_timezone(),
                                                                                      DecimalType.of(9, 2),
                                                                                      StringType.get(),
                                                                                      UUIDType.get(),
                                                                                      FixedType.of_length(1),
                                                                                      BinaryType.get()])

    def test_invalid_time_conversions(self):
        self.assert_invalid_conversions(Literal.of("14:21:01.919")
                                        .to(TimeType.get()), [BooleanType.get(),
                                                              IntegerType.get(),
                                                              LongType.get(),
                                                              FloatType.get(),
                                                              DoubleType.get(),
                                                              DateType.get(),
                                                              TimestampType.with_timezone(),
                                                              TimestampType.without_timezone(),
                                                              DecimalType.of(9, 2),
                                                              StringType.get(),
                                                              UUIDType.get(),
                                                              FixedType.of_length(1),
                                                              BinaryType.get()])

    def test_invalid_timestamp_conversions(self):
        self.assert_invalid_conversions(Literal.of("2017-08-18T14:21:01.919")
                                        .to(TimestampType.without_timezone()), [BooleanType.get(),
                                                                                IntegerType.get(),
                                                                                LongType.get(),
                                                                                FloatType.get(),
                                                                                DoubleType.get(),
                                                                                TimeType.get(),
                                                                                DecimalType.of(9, 2),
                                                                                StringType.get(),
                                                                                UUIDType.get(),
                                                                                FixedType.of_length(1),
                                                                                BinaryType.get()])

    def test_invalid_decimal_conversions(self):
        self.assert_invalid_conversions(Literal.of(Decimal("34.11")), [BooleanType.get(),
                                                                       IntegerType.get(),
                                                                       LongType.get(),
                                                                       FloatType.get(),
                                                                       DoubleType.get(),
                                                                       DateType.get(),
                                                                       TimeType.get(),
                                                                       TimestampType.with_timezone(),
                                                                       TimestampType.without_timezone(),
                                                                       DecimalType.of(9, 4),
                                                                       StringType.get(),
                                                                       UUIDType.get(),
                                                                       FixedType.of_length(1),
                                                                       BinaryType.get()])

    def test_invalid_string_conversions(self):
        self.assert_invalid_conversions(Literal.of("abc"), [BooleanType.get(),
                                                            IntegerType.get(),
                                                            LongType.get(),
                                                            FloatType.get(),
                                                            DoubleType.get(),
                                                            FixedType.of_length(1),
                                                            BinaryType.get()])

    def test_invalid_uuid_conversions(self):
        self.assert_invalid_conversions(Literal.of(uuid.uuid4()), [BooleanType.get(),
                                                                   IntegerType.get(),
                                                                   LongType.get(),
                                                                   FloatType.get(),
                                                                   DoubleType.get(),
                                                                   DateType.get(),
                                                                   TimeType.get(),
                                                                   TimestampType.with_timezone(),
                                                                   TimestampType.without_timezone(),
                                                                   DecimalType.of(9, 2),
                                                                   StringType.get(),
                                                                   FixedType.of_length(1),
                                                                   BinaryType.get()])

    def test_invalid_fixed_conversions(self):
        if sys.version_info[0] >= 3:
            self.assert_invalid_conversions(Literal.of(bytes([0x00, 0x01, 0x02])), [BooleanType.get(),
                                                                                    IntegerType.get(),
                                                                                    LongType.get(),
                                                                                    FloatType.get(),
                                                                                    DoubleType.get(),
                                                                                    DateType.get(),
                                                                                    TimeType.get(),
                                                                                    TimestampType.with_timezone(),
                                                                                    TimestampType.without_timezone(),
                                                                                    DecimalType.of(9, 2),
                                                                                    StringType.get(),
                                                                                    UUIDType.get(),
                                                                                    FixedType.of_length(1)])

    def test_invalid_binary_conversions(self):
        if sys.version_info[0] >= 3:
            self.assert_invalid_conversions(Literal.of(bytearray([0x00, 0x01, 0x02])), [BooleanType.get(),
                                                                                        IntegerType.get(),
                                                                                        LongType.get(),
                                                                                        FloatType.get(),
                                                                                        DoubleType.get(),
                                                                                        DateType.get(),
                                                                                        TimeType.get(),
                                                                                        TimestampType.with_timezone(),
                                                                                        TimestampType.without_timezone(),
                                                                                        DecimalType.of(9, 2),
                                                                                        StringType.get(),
                                                                                        UUIDType.get(),
                                                                                        FixedType.of_length(1)])

    def assert_invalid_conversions(self, lit, types=None):
        for type_var in types:
            self.assertIsNone(lit.to(type_var))

File: api/src/test/java/org/apache/iceberg/expressions/TestMiscLiteralConversions.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.iceberg.expressions;

import static org.assertj.core.api.Assertions.assertThat;

import java.math.BigDecimal;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;
import org.apache.iceberg.types.Type;
import org.apache.iceberg.types.Types;
import org.junit.jupiter.api.Test;

public class TestMiscLiteralConversions {
  @Test
  public void testIdentityConversions() {
    List<Pair<Literal<?>, Type>> pairs =
        Arrays.asList(
            Pair.of(Literal.of(true), Types.BooleanType.get()),
            Pair.of(Literal.of(34), Types.IntegerType.get()),
            Pair.of(Literal.of(34L), Types.LongType.get()),
            Pair.of(Literal.of(34.11F), Types.FloatType.get()),
            Pair.of(Literal.of(34.55D), Types.DoubleType.get()),
            Pair.of(Literal.of("34.55"), Types.DecimalType.of(9, 2)),
            Pair.of(Literal.of("2017-08-18"), Types.DateType.get()),
            Pair.of(Literal.of("14:21:01.919"), Types.TimeType.get()),
            Pair.of(Literal.of("2017-08-18T14:21:01.919432"), Types.TimestampType.withoutZone()),
            Pair.of(
                Literal.of("2017-08-18T14:21:01.919432755"), Types.TimestampNanoType.withoutZone()),
            Pair.of(Literal.of("abc"), Types.StringType.get()),
            Pair.of(Literal.of(UUID.randomUUID()), Types.UUIDType.get()),
            Pair.of(Literal.of(new byte[] {0, 1, 2}), Types.FixedType.ofLength(3)),
            Pair.of(Literal.of(ByteBuffer.wrap(new byte[] {0, 1, 2})), Types.BinaryType.get()));

    for (Pair<Literal<?>, Type> pair : pairs) {
      Literal<?> lit = pair.first();
      Type type = pair.second();

      // first, convert the literal to the target type (date/times start as strings)
      Literal<?> expected = lit.to(type);

      // then check that converting again to the same type results in an identical literal
      assertThat(expected.to(type))
          .as("Converting twice should produce identical values")
          .isSameAs(expected);
    }
  }

  @Test
  public void testTimestampWithMicrosecondsToDate() {
    final Literal<Long> micros =
        Literal.of("2017-08-18T14:21:01.919432755").to(Types.TimestampType.withoutZone());
    final Literal<Integer> dateOfNanos = micros.to(Types.DateType.get());
    assertThat(dateOfNanos).isEqualTo(Literal.of("2017-08-18").to(Types.DateType.get()));
  }

  @Test
  public void testTimestampWithNanoosecondsToDate() {
    final Literal<Long> nanos =
        Literal.of("2017-08-18T14:21:01.919432755").to(Types.TimestampNanoType.withoutZone());
    final Literal<Integer> dateOfNanos = nanos.to(Types.DateType.get());
    assertThat(dateOfNanos).isEqualTo(Literal.of("2017-08-18").to(Types.DateType.get()));
  }

  @Test
  public void testBinaryToFixed() {
    Literal<ByteBuffer> lit = Literal.of(ByteBuffer.wrap(new byte[] {0, 1, 2}));
    Literal<ByteBuffer> fixedLit = lit.to(Types.FixedType.ofLength(3));
    assertThat(fixedLit).as("Should allow conversion to correct fixed length").isNotNull();
    assertThat(fixedLit.value().duplicate())
        .as("Conversion should not change value")
        .isEqualTo(lit.value().duplicate());

    assertThat(lit.to(Types.FixedType.ofLength(4)))
        .as("Should not allow conversion to different fixed length")
        .isNull();
    assertThat(lit.to(Types.FixedType.ofLength(2)))
        .as("Should not allow conversion to different fixed length")
        .isNull();
  }

  @Test
  public void testFixedToBinary() {
    Literal<ByteBuffer> lit = Literal.of(new byte[] {0, 1, 2});
    Literal<ByteBuffer> binaryLit = lit.to(Types.BinaryType.get());
    assertThat(binaryLit).as("Should allow conversion to binary").isNotNull();
    assertThat(binaryLit.value().duplicate())
        .as("Conversion should not change value")
        .isEqualTo(lit.value().duplicate());
  }

  @Test
  public void testInvalidBooleanConversions() {
    testInvalidConversions(
        Literal.of(true),
        Types.IntegerType.get(),
        Types.LongType.get(),
        Types.FloatType.get(),
        Types.DoubleType.get(),
        Types.DateType.get(),
        Types.TimeType.get(),
        Types.TimestampType.withoutZone(),
        Types.TimestampType.withZone(),
        Types.TimestampNanoType.withoutZone(),
        Types.TimestampNanoType.withZone(),
        Types.DecimalType.of(9, 2),
        Types.StringType.get(),
        Types.UUIDType.get(),
        Types.FixedType.ofLength(1),
        Types.BinaryType.get());
  }

  @Test
  public void testInvalidIntegerConversions() {
    testInvalidConversions(
        Literal.of(34),
        Types.BooleanType.get(),
        Types.TimeType.get(),
        Types.TimestampType.withoutZone(),
        Types.TimestampType.withZone(),
        Types.TimestampNanoType.withoutZone(),
        Types.TimestampNanoType.withZone(),
        Types.StringType.get(),
        Types.UUIDType.get(),
        Types.FixedType.ofLength(1),
        Types.BinaryType.get());
  }

  @Test
  public void testInvalidLongConversions() {
    testInvalidConversions(
        Literal.of(34L),
        Types.BooleanType.get(),
        Types.StringType.get(),
        Types.UUIDType.get(),
        Types.FixedType.ofLength(1),
        Types.BinaryType.get());
  }

  @Test
  public void testInvalidFloatConversions() {
    testInvalidConversions(
        Literal.of(34.11F),
        Types.BooleanType.get(),
        Types.IntegerType.get(),
        Types.LongType.get(),
        Types.DateType.get(),
        Types.TimeType.get(),
        Types.TimestampType.withoutZone(),
        Types.TimestampType.withZone(),
        Types.TimestampNanoType.withoutZone(),
        Types.TimestampNanoType.withZone(),
        Types.StringType.get(),
        Types.UUIDType.get(),
        Types.FixedType.ofLength(1),
        Types.BinaryType.get());
  }

  @Test
  public void testInvalidDoubleConversions() {
    testInvalidConversions(
        Literal.of(34.11D),
        Types.BooleanType.get(),
        Types.IntegerType.get(),
        Types.LongType.get(),
        Types.DateType.get(),
        Types.TimeType.get(),
        Types.TimestampType.withoutZone(),
        Types.TimestampType.withZone(),
        Types.TimestampNanoType.withoutZone(),
        Types.TimestampNanoType.withZone(),
        Types.StringType.get(),
        Types.UUIDType.get(),
        Types.FixedType.ofLength(1),
        Types.BinaryType.get());
  }

  @Test
  public void testInvalidDateConversions() {
    testInvalidConversions(
        Literal.of("2017-08-18").to(Types.DateType.get()),
        Types.BooleanType.get(),
        Types.IntegerType.get(),
        Types.LongType.get(),
        Types.FloatType.get(),
        Types.DoubleType.get(),
        Types.TimeType.get(),
        Types.TimestampType.withoutZone(),
        Types.TimestampType.withZone(),
        Types.TimestampNanoType.withoutZone(),
        Types.TimestampNanoType.withZone(),
        Types.DecimalType.of(9, 4),
        Types.StringType.get(),
        Types.UUIDType.get(),
        Types.FixedType.ofLength(1),
        Types.BinaryType.get());
  }

  @Test
  public void testInvalidTimeConversions() {
    testInvalidConversions(
        Literal.of("14:21:01.919").to(Types.TimeType.get()),
        Types.BooleanType.get(),
        Types.IntegerType.get(),
        Types.LongType.get(),
        Types.FloatType.get(),
        Types.DoubleType.get(),
        Types.DateType.get(),
        Types.TimestampType.withoutZone(),
        Types.TimestampType.withZone(),
        Types.TimestampNanoType.withoutZone(),
        Types.TimestampNanoType.withZone(),
        Types.DecimalType.of(9, 4),
        Types.StringType.get(),
        Types.UUIDType.get(),
        Types.FixedType.ofLength(1),
        Types.BinaryType.get());
  }

  @Test
  public void testInvalidTimestampMicrosConversions() {
    testInvalidConversions(
        Literal.of("2017-08-18T14:21:01.919123").to(Types.TimestampType.withoutZone()),
        Types.BooleanType.get(),
        Types.IntegerType.get(),
        Types.LongType.get(),
        Types.FloatType.get(),
        Types.DoubleType.get(),
        Types.TimeType.get(),
        Types.DecimalType.of(9, 4),
        Types.StringType.get(),
        Types.UUIDType.get(),
        Types.FixedType.ofLength(1),
        Types.BinaryType.get());
  }

  @Test
  public void testInvalidTimestampNanosConversions() {
    testInvalidConversions(
        Literal.of("2017-08-18T14:21:01.919123456").to(Types.TimestampNanoType.withoutZone()),
        Types.BooleanType.get(),
        Types.IntegerType.get(),
        Types.LongType.get(),
        Types.FloatType.get(),
        Types.DoubleType.get(),
        Types.TimeType.get(),
        Types.DecimalType.of(9, 4),
        Types.StringType.get(),
        Types.UUIDType.get(),
        Types.FixedType.ofLength(1),
        Types.BinaryType.get());
  }

  @Test
  public void testInvalidDecimalConversions() {
    testInvalidConversions(
        Literal.of(new BigDecimal("34.11")),
        Types.BooleanType.get(),
        Types.IntegerType.get(),
        Types.LongType.get(),
        Types.FloatType.get(),
        Types.DoubleType.get(),
        Types.DateType.get(),
        Types.TimeType.get(),
        Types.TimestampType.withoutZone(),
        Types.TimestampType.withZone(),
        Types.TimestampNanoType.withoutZone(),
        Types.TimestampNanoType.withZone(),
        Types.StringType.get(),
        Types.UUIDType.get(),
        Types.FixedType.ofLength(1),
        Types.BinaryType.get());
  }

  @Test
  public void testInvalidStringConversions() {
    // Strings can be used for types that are difficult to construct, like decimal or timestamp,
    // but are not intended to support parsing strings to any type
    testInvalidConversions(
        Literal.of("abc"),
        Types.BooleanType.get(),
        Types.IntegerType.get(),
        Types.LongType.get(),
        Types.FloatType.get(),
        Types.DoubleType.get(),
        Types.FixedType.ofLength(1),
        Types.BinaryType.get());
  }

  @Test
  public void testInvalidUUIDConversions() {
    testInvalidConversions(
        Literal.of(UUID.randomUUID()),
        Types.BooleanType.get(),
        Types.IntegerType.get(),
        Types.LongType.get(),
        Types.FloatType.get(),
        Types.DoubleType.get(),
        Types.DateType.get(),
        Types.TimeType.get(),
        Types.TimestampType.withoutZone(),
        Types.TimestampType.withZone(),
        Types.TimestampNanoType.withoutZone(),
        Types.TimestampNanoType.withZone(),
        Types.DecimalType.of(9, 2),
        Types.StringType.get(),
        Types.FixedType.ofLength(1),
        Types.BinaryType.get());
  }

  @Test
  public void testInvalidFixedConversions() {
    testInvalidConversions(
        Literal.of(new byte[] {0, 1, 2}),
        Types.BooleanType.get(),
        Types.IntegerType.get(),
        Types.LongType.get(),
        Types.FloatType.get(),
        Types.DoubleType.get(),
        Types.DateType.get(),
        Types.TimeType.get(),
        Types.TimestampType.withoutZone(),
        Types.TimestampType.withZone(),
        Types.TimestampNanoType.withoutZone(),
        Types.TimestampNanoType.withZone(),
        Types.DecimalType.of(9, 2),
        Types.StringType.get(),
        Types.UUIDType.get(),
        Types.FixedType.ofLength(1));
  }

  @Test
  public void testInvalidBinaryConversions() {
    testInvalidConversions(
        Literal.of(ByteBuffer.wrap(new byte[] {0, 1, 2})),
        Types.BooleanType.get(),
        Types.IntegerType.get(),
        Types.LongType.get(),
        Types.FloatType.get(),
        Types.DoubleType.get(),
        Types.DateType.get(),
        Types.TimeType.get(),
        Types.TimestampType.withoutZone(),
        Types.TimestampType.withZone(),
        Types.TimestampNanoType.withoutZone(),
        Types.TimestampNanoType.withZone(),
        Types.DecimalType.of(9, 2),
        Types.StringType.get(),
        Types.UUIDType.get(),
        Types.FixedType.ofLength(1));
  }

  private void testInvalidConversions(Literal<?> lit, Type... invalidTypes) {
    for (Type type : invalidTypes) {
      assertThat(lit.to(type))
          .as(lit.value().getClass().getName() + " literal to " + type + " is not allowed")
          .isNull();
    }
  }

  private static class Pair<X, Y> {
    public static <X, Y> Pair<X, Y> of(X first, Y second) {
      return new Pair<>(first, second);
    }

    private final X first;
    private final Y second;

    private Pair(X first, Y second) {
      this.first = first;
      this.second = second;
    }

    public X first() {
      return first;
    }

    public Y second() {
      return second;
    }
  }
}
